src/raphael_backend_flask/app.py

```diff
@@ -0,0 +1,255 @@
+from dotenv import load_dotenv
+import requests
+
+load_dotenv()
+
+import os
+import json
+from typing import Any
+import sqlite3
+from sqlite3 import Connection, Row
+from urllib.parse import urlparse, parse_qs
+
+from flask import Flask, redirect, request, jsonify, g, url_for
+from flask.typing import ResponseReturnValue
+from flask_httpauth import HTTPBasicAuth
+from flask_cors import CORS
+from werkzeug.security import generate_password_hash, check_password_hash
+
+
+from db import create_database
+from vertex import process_video
+from youtube import download_captions, extract_title
+
+DB_PATH = os.getenv("DB_PATH", "database.db")
+
+app = Flask(__name__)
+CORS(app)
+
+auth = HTTPBasicAuth()
+
+users = {
+    user: generate_password_hash(password)
+    for user, password in (item.split(":") for item in os.environ["USERS"].split(","))
+}
+
+
+@auth.verify_password
+def verify_password(username: str, password: str) -> str | None:
+    if username in users and check_password_hash(users[username], password):
+        return username
+
+
+def get_db_connection() -> Connection:
+    conn = g.get("_database")
+    if not conn:
+        if not os.path.isfile(DB_PATH):
+            create_database(DB_PATH)
+        conn = g._database = sqlite3.connect(f"file:{DB_PATH}?mode=rw", uri=True)
+    conn.row_factory = Row
+    return conn
+
+
+@app.teardown_appcontext
+def teardown_db_connection(_: Any) -> None:
+    conn: Connection | None = g.get("_database")
+    if conn:
+        conn.close()
+
+
+def execute_sql(sql: str, params: tuple[Any, ...] = ()) -> list[Row]:
+    """Open/close the database for each request, we don't expect to make many."""
+    conn = get_db_connection()
+    cur = conn.cursor()
+    cur.execute("PRAGMA foreign_keys = ON")
+    cur.execute(sql, params)
+    data = cur.fetchall()
+    conn.commit()
+    cur.close()
+    return data
+
+
+@app.get("/api/transcripts/")
+@auth.login_required
+def get_transcripts() -> ResponseReturnValue:
+    transcripts = execute_sql("SELECT * FROM video_transcripts")
+    return jsonify([{**t} for t in transcripts]), 200
+
+
+def extract_youtube_id(url: str) -> str:
+    try:
+        parsed = urlparse(url)
+        if parsed.netloc == "youtu.be":
+            return parsed.path[1:]
+
+        if parsed.netloc == "" and len(parsed.path) == 11:
+            # if it's 11 chars like a video id, wing it
+            return parsed.path
+
+        queries = parse_qs(parsed.query)
+        # assume this exists
+        return queries["v"][0]
+    except:
+        raise Exception
```

Comment:
I’d imagined using a regex here, but I think what you’ve done is nicer.

Two (hopefully small) things though:

1. Please avoid using `except: raise Exception`. It’s preferable to catch a specific issue (in this case, I think we can expect to see `KeyError`s and `IndexError`s), and raise a clear issue (either a custom exception, or an exception with a message). You could do `queries.get("v", [])[0]` and then I think we only expect `IndexError`s.

2. Minor thing, and this is more a suggestion because I think what you’ve done is fine: You could consider using `if parsed.netloc in ["m.youtube.com", "www.youtube.com", "youtube.com]:`. This will prevent things like http://example.com?v=dQw4w9WgXcQ being accepted.

Reply:
This is really just for the mvp out the door. If you're putting in non-YouTube URLs you're intentionally putting garbage in and getting garbage out. I would enumerate whitelisted netlocs if it felt like a big deal. Same for the generic exception handler.